### PR TITLE
[improv] using proxy for http request

### PR DESCRIPTION
**Pull request by msimon**
This commit add proxy support to the agent. It's configure thanks the config file with 3 required parameter: "proxy-type", "proxy-url", "proxy-port".

We have been testing it in LinkedIn for a couple of weeks and it's seems to work correctly.

Thanks.

### CODE REVIEW

File: src/le.py [hunk 1]
```diff
@@ -1700,6 +1705,27 @@ def __init__(self, endpoint, port, use_ssl, preamble, debug_transport_events):
 
         self._shutdown = False
 
+        # proxy setup
+        self._use_proxy = False
+
+        (proxy_type_str, self._proxy_url, self._proxy_port) = proxy
+
+        if (proxy_type_str != NOT_SET and self._proxy_url != NOT_SET and self._proxy_port != NOT_SET):
```
Comment (vilda) [inline src/le.py]: Note multiple instances of C-ism ;) Brackets are not needed inside if statements

File: src/le.py [hunk 2]
```diff
@@ -1789,7 +1818,13 @@ def _open_connection(self):
         while not self._shutdown:
             self._close_connection()
             try:
-                s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                s = None
+                if (self._use_proxy == True):
```
Comment (vilda) [inline src/le.py]: Python way is ```if self._use_proxy:```

File: src/le.py [hunk 3]
```diff
@@ -2081,6 +2123,9 @@ def load(self, load_include_dirs=True):
                 V1_METRICS_PARAM: 'True',
                 PULL_SERVER_SIDE_CONFIG_PARAM: 'True',
                 INCLUDE_PARAM: '',
+                PROXY_TYPE_PARAM: '',
+                PROXY_URL_PARAM: '',
+                PROXY_PORT_PARAM: ''
```
Comment (vilda) [inline src/le.py]: Usually we close lists definitions with comma at the end

File: src/le.py [hunk 4]
```diff
@@ -3549,7 +3616,7 @@ def main_root():
         # Filesystem operations
         'ls': cmd_ls,
         'rm': cmd_rm,
-        'pull': cmd_pull,
+        'pull': cmd_pull
```
Comment (vilda) [inline src/le.py]: Ah, apparently you consistently don't like commas :)

File: src/utils.py [hunk 5]
```diff
@@ -254,14 +275,25 @@ def domain_connect(config, domain, Domain):
...
         else:
             port = 8081
-    s = '%s:%s' % (s, port)
-    log.debug('Connecting to %s', s)
+    log.debug('Connecting to %s:%s', s, port)
 
     # Pass the connection
     if use_ssl:
-        return make_https_connection(config, s)
+        print "Using SSL"
```
Comment (vilda) [inline src/utils.py]: Leftover debug message?

File: src/utils.py [hunk 6]
```diff
@@ -254,14 +275,25 @@ def domain_connect(config, domain, Domain):
...
+        return make_https_connection(config, s, port)
     else:
-        return httplib.HTTPConnection(s)
+        if config.use_proxy == True:
+            print "Using Proxy"
```
Comment (vilda) [inline src/utils.py]: Another one

File: src/utils.py [hunk 7]
```diff
@@ -254,14 +275,25 @@ def domain_connect(config, domain, Domain):
...
+            conn = httplib.HTTPConnection(config.proxy_url, config.proxy_port)
+            if hasattr(httplib.HTTPConnection, "set_tunnel"):
+                conn.set_tunnel(s, port)
+            else:
+                conn._set_tunnel(s, port)
+            return conn
+        else:
+            print "Not using SSL not proxy"
```
Comment (vilda) [inline src/utils.py]: And another one

**Comment by vilda** [comment]
Thanks @msimon great work!

If you get chance to look at small issues around the code. NP if not, I can do it myself later. I'll also need to check the license and update licensing information.

You can add yourself to contributors (or I can do it) if that's fine with you?

**Comment by msimon** [comment]
@vilda Haha yeah this is actually the only piece of code I wrote in python.
Fixing all of that today.

**Comment by udangel-r7** [comment]
One small comment, can't we join port and host together? and what about support HTTP_PROXY/HTTPS_PROXY environment variables? Would that be of interest?

**Comment by msimon** [comment]
@udangel-r7 Looking at HTTPConnection.set_tunnel documentation it's unclear if the port can be combined to the uri. Since I was not sure I played it safe, if you know for sure it is supported I can update the code.
- Environment variables could be interesting for some, it is not for us. If you have an implementation sketch I can also implement it.

**Comment by udangel-r7** [comment]
Sorry for the big delay, is there a update on this? @msimon should we do the changes?

**Comment by msimon** [comment]
Sorry about this, I have a lot of work lately.
You can go ahead and do the change, or I will do them next week (starting 29th) when I have more time on my hand.

Lmk

**Comment by msimon** [comment]
@vilda @udangel-r7 Finally had time to edit the commit. Fix all of the issues that you guys raised.

Let me know if you see anything else.

Thanks!

**Comment by msimon** [comment]
@vilda @udangel-r7 around to check this ?

Thanks

**Comment by vilda** [comment]
Sorry for the delay @msimon will definitely have a look, this is great

**Pull request merged**